draft-ietf-mls-protocol.md

```diff
@@ -4488,16 +4471,23 @@ state compromise.
 
 ## Delayed and Reordered Application messages
 
-Since each Application message contains the group identifier, the epoch and a
-message counter, a client can receive messages out of order.
-If they are able to retrieve or recompute the correct AEAD decryption key
-from currently stored cryptographic material clients can decrypt
-these messages.
-
-For usability, MLS clients might be required to keep the AEAD key
-and nonce for a certain amount of time to retain the ability to decrypt
-delayed or out of order messages, possibly still in transit while a
-decryption is being done.
+Since each application message contains the group identifier, the epoch, and a
+generation counter, a client can receive messages out of order. When messages
+are received out of order, the client moves the sender ratchet forward to match
+the received generation counter. Any unused nonce and key pairs from the ratchet
+are potentially stored so that they can be used to decrypt the messages which
+were delayed or reordered.
+
+Applications SHOULD define a policy on how long to keep unused nonce and key
+pairs for a sender, and the maximum number to keep. This is in addition to
+ensuring that these nonce and key pairs are promptly deleted when the epoch
+ends. Applications SHOULD also define a policy limiting the maximum number of
+steps that clients will move the ratchet forward in response to a new message. Messages received
+with an unusually high generation counter would then be rejected to avoid
```

Comment:
"unusually high" is a bit vague. Maybe we can be more specific here by saying that clients can enforce a maximum difference for the generation counter between two subsequent messages.